pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -1013,6 +1013,47 @@ public boolean tryLoadExistingSegment(SegmentZKMetadata zkMetadata, IndexLoading
     }
   }
 
+  @Override
+  public boolean checkReloadSegment(SegmentZKMetadata zkMetadata, IndexLoadingConfig indexLoadingConfig) {
+    String segmentName = zkMetadata.getSegmentName();
+
+    // Creates the SegmentDirectory object to access the segment metadata.
+    // The metadata is null if the segment doesn't exist yet.
+    SegmentDirectory segmentDirectory =
+        tryInitSegmentDirectory(segmentName, String.valueOf(zkMetadata.getCrc()), indexLoadingConfig);
+    SegmentMetadataImpl segmentMetadata = (segmentDirectory == null) ? null : segmentDirectory.getSegmentMetadata();
+
+    // If the segment doesn't exist on server or its CRC has changed, then we
+    // need to fall back to download the segment from deep store to load it.
+    if (segmentMetadata == null || !hasSameCRC(zkMetadata, segmentMetadata)) {
+      if (segmentMetadata == null) {
+        _logger.info("Segment: {} does not exist", segmentName);
+      } else if (!hasSameCRC(zkMetadata, segmentMetadata)) {
+        _logger.info("Segment: {} has CRC changed from: {} to: {}", segmentName, segmentMetadata.getCrc(),
+            zkMetadata.getCrc());
+      }
+      closeSegmentDirectoryQuietly(segmentDirectory);
+      return true;
+    }
+
+    try {
+      Schema schema = indexLoadingConfig.getSchema();
+      //if the reload of the segment is not needed then return false
+      if (!ImmutableSegmentLoader.needPreprocess(segmentDirectory, indexLoadingConfig, schema)) {
+        _logger.info("Segment: {} is consistent with latest table config and schema", segmentName);
```

Comment:
This can flood the log (consider thousands of segments on each server)

Reply:
removed the log